fix(bingo): correct Windows paths handling

astrochemx opened the pull request:
<!-- 👋 Hi, thanks for sending a PR to Bingo! 💖.
Please fill out all fields below and make sure each item is true and [x] checked.
Otherwise we may not be able to review your PR. -->

## PR Checklist

- [x] Addresses an existing open issue: fixes #300
- [x] That issue was marked as [`status: accepting prs`](https://github.com/JoshuaKGoldberg/bingo/issues?q=is%3Aopen+is%3Aissue+label%3A%22status%3A+accepting+prs%22)
- [x] Steps in [CONTRIBUTING.md](https://github.com/JoshuaKGoldberg/bingo/blob/main/.github/CONTRIBUTING.md) were taken

## Overview

<!-- Description of what is changed and how the code change does that. -->

### :monocle_face: The problem

On Windows, file URLs in the `Node.js` environment often include an extra slash after the `file://` protocol, resulting in URLs like `file:///C:/Users/User/file.js`. When parsed using `.pathname`, the output includes a leading slash before the drive letter, yielding `/C:/Users/User/file.js`. While this leading slash complies with the URL specification to denote an absolute path, Windows uses drive letters instead of a root directory. Consequently, this results in an invalid file path since Windows paths should begin directly with the drive letter, such as `C:/Users/User/file.js`. Moreover, mishandling paths with a leading slash can lead to misinterpretation. For example, paths like `/C:/Users/User/file.js` may be treated as relative to the current drive, potentially producing an erroneous path with duplicated drive letters like `C:/C:/Users/User/file.js`. As for the use of forward and backward slashes on Windows, I encountered no issues in this context, since Windows generally accepts both types of slashes.

### :bulb: The solution

I see two possible approaches to solve this problem, each with its advantages.

---

**1. Using a Regular Expressions for Platform-Specific Path Fixing**

This is the approach I used as a base for my PR.
It removes any leading slashes in the absolute file paths on Windows platforms using a simple RegEx:

```ts
function fixPathIfWindows(path: string) {
	return process.platform === "win32"
		? path.replace(/^[/\\]+([A-Z]:[/\\])/i, "$1")
		: path;
}

const directory = fixPathIfWindows(path.dirname(callId.file));
```

#### Why?
- Lightweight and straightforward.
- Avoids unnecessary module imports.
- Sufficient for resolving the current issue without introducing additional complexity.

---

**2. Using `fileURLToPath` for Cross-Platform Compatibility**

Alternatively, `fileURLToPath` from the `node:url` module can be used to reliably convert a file URL to a native path.
Combined with `new URL(path, 'file://')` for URL construction, this provides robust, cross-platform compatibility:

```ts
import { dirname } from "path";
import { fileURLToPath } from "url";

const directory = dirname(fileURLToPath(new URL(callId.file, 'file://')));
```

#### Why?
- Ensures correct path conversion on all platforms without relying on manual string manipulation.
- Built-in `Node.js` utility specifically designed for handling file URLs safely.

---

Both approaches are valid, but since the first one directly addresses the issue with minimal overhead, I opted for it in my PR.

Looking forward to your review and feedback.
Thanks!

JoshuaKGoldberg commented on packages/bingo/src/cli/utils.ts:
```diff
@@ -1,3 +1,9 @@
+export function fixPathIfWindows(path: string) {
+	return process.platform === "win32"
+		? path.replace(/^[/\\]+([A-Z]:[/\\])/i, "$1")
+		: path;
```
> 2. Using `fileURLToPath` for Cross-Platform Compatibility

IMO we should go with that one. The less custom code in the project the better. Especially code that deals with path intricacies. _Especially_ regular expressions. 😄 

If Node.js has already handled something for us then let's go with that.

One quick note: a stylistic preference for `node:url` over `url`.

JoshuaKGoldberg reviewed (changes requested):
Thanks so much for the detailed & quick PR! Super helpful in figuring out what's going on. Agreed on your analysis in general. Since `fileURLToPath` is already written for us I'd say let's go for it.

Could you also add a unit test for this please? `getTemplatePackageData.test.ts` already exists and should be well-positioned to handle the new case.

astrochemx committed:
fix(bingo): correct Windows paths handling

astrochemx commented:
Hi, Josh!

I've corrected the path handling approach and force-pushed the updated branch.

UPD: Since [`Node.js` v22.1.0 and v20.13.0](https://nodejs.org/api/url.html#urlfileurltopathurl-options), `fileURLToPath` supports an `options` argument that allows us to specify how to parse the path - either as a Windows or POSIX file path. This means I might be able to write platform-specific tests that pass consistently on any platform.

UPD: I've added some new tests, modified the existing ones to properly test how the new `resolveFilePath` method handles paths on different platforms, and removed a duplicated test.

UPD: Should I also update `import path from "path"` to `import path from "node:path"` or even `import { dirname } from "node:path"` while we're at it?

I look forward to your comments and suggestions.
Thanks!

PS: Also, the same error resulting in a double drive letter, such as `C:\C:\GHub\create-typescript-app\node_modules\.pnpm\cspell-populate-words@0.3.0\node_modules\cspell-populate-words\bin\index.mjs`, occurs on Windows in the [`resolveBin` function](https://github.com/JoshuaKGoldberg/create-typescript-app/blob/main/src/utils/resolveBin.ts) of the `create-typescript-app`. Should I [address this issue](https://github.com/astrochemx/create-typescript-app/commit/7c2b3cd781e7b9177d8f97442ad3ad0bf567c8f8) as well after we solve this one?

<details>
<summary>Example of an error:</summary>

```
Error: Cannot find module 'C:\C:\GHub\create-typescript-app\node_modules\.pnpm\cspell-populate-words@0.3.0\node_modules\cspell-populate-words\bin\index.mjs'
    at Function._resolveFilename (node:internal/modules/cjs/loader:1405:15)
    at defaultResolveImpl (node:internal/modules/cjs/loader:1061:19)
    at resolveForCJSWithHooks (node:internal/modules/cjs/loader:1066:22)
    at Function._load (node:internal/modules/cjs/loader:1215:37)
    at TracingChannel.traceSync (node:diagnostics_channel:322:14)
    at wrapModuleLoad (node:internal/modules/cjs/loader:235:24)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:151:5)
    at node:internal/main/run_main_module:33:47 {
      code: 'MODULE_NOT_FOUND',
      requireStack: []
    }
Node.js v23.10.0
    at getFinalError (file:///C:/GHub/bingo/node_modules/.pnpm/execa@9.5.2/node_modules/execa/lib/return/final-error.js:6:9)
    at makeError (file:///C:/GHub/bingo/node_modules/.pnpm/execa@9.5.2/node_modules/execa/lib/return/result.js:108:16)
    at getAsyncResult (file:///C:/GHub/bingo/node_modules/.pnpm/execa@9.5.2/node_modules/execa/lib/methods/main-async.js:167:4)       
    at handlePromise (file:///C:/GHub/bingo/node_modules/.pnpm/execa@9.5.2/node_modules/execa/lib/methods/main-async.js:150:17)       
    at async runCommand (file:///C:/GHub/bingo/packages/bingo/lib/runners/applyScriptsToSystem.js:10:24)
    at async file:///C:/GHub/bingo/packages/bingo/lib/runners/applyScriptsToSystem.js:22:17
    at async Promise.all (index 1)
    at async applyScriptsToSystem (file:///C:/GHub/bingo/packages/bingo/lib/runners/applyScriptsToSystem.js:20:9)
    at async Promise.all (index 0)
    at async runCreation (file:///C:/GHub/bingo/packages/bingo/lib/runners/runCreation.js:8:5)
```

</details>

astrochemx committed:
test(bingo): enhance `getTemplatePackageData` tests for Windows and POSIX paths

JoshuaKGoldberg commented on packages/bingo/src/cli/getTemplatePackageData.test.ts:
```diff
@@ -18,6 +19,32 @@ vi.mock("read-package-up", () => ({
+	};
+});
+
+vi.mock("./utils.ts", { spy: true });
```
[Testing] 🤔 I don't see much benefit to testing these internals. As long as the result from `getTemplatePackageData` respects Windows, I don't think we want to care about what happened internally. Is there a reason to include this?

astrochemx replied:
Hi, Josh!
Thanks for your feedback and sorry for the delay!

The test was failing because I needed to mock `node:path` in addition to `node:url` to ensure platform-specific behavior. I've addressed that now. And I think you're right - in this case, there’s not much benefit in testing the internals, so I’ve refactored and simplified the tests to ensure they pass on both `POSIX` (tested on an Ubuntu WSL) and `Windows` platforms.

Please let me know if you have any further suggestions.
I look forward to your feedback, thanks!

JoshuaKGoldberg reviewed (changes requested):
Just a high-level question, looks like there are some test failures in CI. I can take a deeper look if tests are passing or there's some blocker to getting them there. Thanks!

astrochemx committed:
test(bingo): refactor tests for `getTemplatePackageData`

JoshuaKGoldberg commented on packages/bingo/src/cli/getTemplatePackageData.test.ts:
```diff
@@ -18,6 +18,56 @@ vi.mock("read-package-up", () => ({
+		output: "/home/user/project",
+	},
+	windows: {
+		input: "/C:/Users/User/file.js",
```
[Question] Should we also test the case of `"C:/Users/User/file.js"`? That's another possible path, right?

astrochemx replied:
Yes, if we plan to use our function with inputs beyond the output from the `call-id` package, then other possible paths on Windows could include (to the best of my knowledge):

- `"C:/Users/User/file.js"`
- `"C:\Users\User\file.js"`

I can create tests for these two cases as well if needed

JoshuaKGoldberg replied:
Sorry, I'm not parsing - are you saying that `call-id` only ever returns `/`-prefixed paths on Windows? Is that a bug in `call-id`? Can you say more?

astrochemx replied:
Just to clarify, I'm not very familiar with `Node.js`.
However, from what I understand, this isn't a bug in `call-id`.
It seems that the package doesn't perform platform-specific path normalization, which is likely why we're seeing `/`-prefixed paths on Windows.

JoshuaKGoldberg replied:
Heh, and I'm not familiar enough with Windows to know whether that's a bug. I should be able to test this PR out on a Windows device later today and can post back.

astrochemx replied:
I was thinking that we may not need to test multiple legitimate Windows paths, since that would essentially be testing the internal behavior of `Node.js`, which we're not interested in. One test with a potentially invalid path for Windows that could occur in our scenario should be sufficient to test path handling through `fileURLToPath` on Windows. Additionally, I believe we won't encounter such legitimate Windows paths in the context where our function is used (with the `call-id` package).

I'd love to hear your opinion on this

astrochemx replied:
I have conducted further investigation into the handling of Windows file paths in `Node.js`.

We resolve paths using `fileURLToPath`, and the input to this function is a `URL` object:
```ts
export function resolveFilePath(filePath: string) {
  return fileURLToPath(new URL(filePath, 'file://'));
}
```

On `Windows`, file paths typically begin with a drive letter (e.g., `C:`). However, for these paths to conform to the `file://` URI scheme in `Node.js`, they must include a leading slash (e.g., `/C:`). As a result, we should end up with a `URL` like: `file:///C:/Users/User/file.js`, and not like `"file://C:/Users/User/file.js"`.

Therefore, **valid paths** in `Node.js` on `Windows` for passing into the `URL` constructor should appear as:
* `"/C:/Users/User/file.js"` (this is the path provided by the `call-id` package, which is the correct format);
* `"\C:\Users\User\file.js"` (although the `URL` constructor accepts this case in my experience, it's unlikely we would encounter a path formatted this way in practice, because `Node.js` typically uses forward slashes (`/`) in file paths when working with the `file://` scheme, and backslashes are not URL-compliant in general).

On the other hand, these paths, which are **valid on Windows**, would be **invalid** if provided to our `resolveFilePath` function:
* `"C:/Users/User/file.js"`;
* `"C:\Users\User\file.js"`

Thus, I'm uncertain about the appropriate course of action. Should I modify our function to prepend the leading slash if the platform is Windows and the provided path starts with a drive letter? However, I don’t think this modification is necessary for our use case, which is why we simplified it in the first place – to do only what we need.

JoshuaKGoldberg commented on packages/bingo/src/cli/getTemplatePackageData.test.ts:
```diff
@@ -31,39 +81,44 @@ describe("getTemplatePackageData", () => {
-				new Error("Could not find a package.json relative to '/path/to'."),
-			);
-		});
+		isWindowsPaths = false;
```
[Testing] Bug: if the test errors, it won't reach this line. So other tests will have a potentially wrong `isWindowsPaths`.

Any sort of test environment cleanup should happen in a `beforeEach` or `afterEach` to make sure subsequent tests aren't tainted.

In this case, I think a `beforeEach` to set `isWindowsPaths` to `false` would make sure tests only opt into it themselves, not accidentally by a previously failed one:

```ts
beforeEach(() => {
  isWindowsPaths = false;
});
```

astrochemx replied:
Yeah, thank you! I hadn't thought of that case.

I'll add:

```ts
beforeEach(() => {
  isWindowsPaths = false;
  vi.resetAllMocks();
});
```

Because otherwise it won't work, we have to reset the mocks

JoshuaKGoldberg commented on packages/bingo/src/cli/getTemplatePackageData.test.ts:
```diff
@@ -18,6 +18,56 @@ vi.mock("read-package-up", () => ({
+			...nodeUrl,
+			fileURLToPath: (url: string | URL) =>
+				nodeUrl.fileURLToPath(url, { windows: isWindowsPaths }),
+		},
```
[Cleanup] Tests still pass if we don't support the `default` export, as no code under test uses it:

```suggestion
```

The opposite with `node:path` mocking, amusingly: it's only ever accessed by the `default` export in code under test.

Requesting: remove any code that isn't necessary for tests to pass.

astrochemx replied:
I have removed the unnecessary code

JoshuaKGoldberg commented on packages/bingo/src/cli/utils.ts:
```diff
@@ -1,3 +1,13 @@
+export function resolveFilePath(filePath: string) {
+	const fileUrl = filePath.startsWith("file://")
+		? new URL(filePath)
+		: new URL(filePath, "file://");
```
[Tests] Nothing exercises the `startsWith("file://")` case right now. If we switch this whole function to:

```ts
export function resolveFilePath(filePath: string) {
	return fileURLToPath(new URL(filePath, "file://"));
}
```

...then tests still pass.

Either:

* Starting with `file://` is a necessary case to handle, and we're missing a test
* It's not a necessary case, so we can trim down this function some

astrochemx replied:
In our case, I don't think handling the `file://` case is necessary, so we could simplify the function for now.
I initially wrote it this way to cover both possible cases for completeness.

Would you prefer me to simplify the function or add a new test case for this scenario?

JoshuaKGoldberg replied:
If it can't be reached by users, then 🔪 let's get rid of it. The less code the better.

JoshuaKGoldberg reviewed (changes requested):
Progress! 🚀

astrochemx committed:
test(bingo): correct tests for `getTemplatePackageData`

astrochemx committed:
refactor(bingo): simplify `resolveFilePath`

astrochemx commented:
Hello, Josh!

Sorry for the confusion earlier.
I took a closer look at the `read-package-up` package and found that it supports using a `URL` as an argument.
With that in mind, we can simplify the code like this:

```ts
const directory = new URL(path.dirname(callId.file), "file://");
const result = await readPackageUp({ cwd: directory });
```

Let me know what you think.
Thanks!

astrochemx commented:
Hi, Josh!

After I took a closer look at the `read-package-up` package and found that it supports using a `URL` object as an argument, I created [another branch](https://github.com/astrochemx/bingo/tree/fix/windows-paths-url) with a slightly different approach.

I ended up changing `getTemplatePackageData.ts` as follows:

```ts
import { getCallId } from "call-id";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { readPackageUp } from "read-package-up";

export async function getTemplatePackageData() {
	const callId = getCallId(2);

	if (!callId) {
		return new Error(
			"Could not determine what directory this Bingo CLI is being called from.",
		);
	}

	const directoryUrl = new URL(path.dirname(callId.file), "file://");
	const result = await readPackageUp({ cwd: directoryUrl });

	return (
		result?.packageJson ??
		new Error(
			`Could not find a package.json relative to '${fileURLToPath(directoryUrl)}'.`,
		)
	);
}
```

This is a somewhat simpler approach without creating a new function.
I'd love to hear your thoughts on it.

I tested both approaches on Windows and WSL (Ubuntu) with [these changes](https://github.com/astrochemx/create-typescript-app/tree/fix/windows-paths) to `create-typescript-app`, and I was able to successfully create an app.

If you prefer this more, I can create a new PR or change this one.

JoshuaKGoldberg commented:
Great investigation! The newer approach seems simpler to me, so I'd say 👍 go for it in this PR.

My old Windows dev laptop wasn't ready to develop on yet the way I thought it was, so I'm not going to be able to fully review until next week. Sorry for the delay. But once you've got it working the way you want it, just post here and I'll see the update.

Thanks!

astrochemx committed:
refactor(bingo): refine handling of caller path in `getTemplatePackageData`

astrochemx commented:
Thank you for your feedback, Josh!

I’ve updated the code and believe these should be the final changes.
If anything else arises, I’ll keep you posted.

I'm looking forward to your review whenever you're ready.
Let me know if you need anything further from my side!

Thanks! 😊

**UPD:** Since this falls within the scope of these changes, I’ve also updated the remaining tests that were previously failing on Windows (`runners/runInput.test.ts` in `bingo` and `intake.test.ts` in `bingo-fs`). I hope my approach to fixing the tests looks good to you. Happy to make any refinements if needed!

**UPD:** Also, maybe we should add a CI testing workflow for Windows?

<details>
<summary>Details</summary>

```yaml
  test_windows:
    name: Test (Windows)
    runs-on: windows-latest
    steps:
      - uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683 # v4
      - uses: ./.github/actions/prepare
      - run: pnpm run -r prebuild
      - run: pnpm run build
      - run: pnpm run test --coverage --test-timeout=20000
      - env:
          CODECOV_TOKEN: ${{ secrets.CODECOV_TOKEN }}
        if: always()
        uses: codecov/codecov-action@v3
```

We should set a longer timeout because on Windows some tests can take up to 20 seconds to pass.

I'm attaching [the last CI workflow run results](https://github.com/astrochemx/bingo/actions/runs/14149326080/job/39640531361) with lower execution times, but I’ve seen durations go up to 18 seconds:

- ✓  bingo  src/cli/runTemplateCLI.test.ts (4 tests) 12311ms
     - ✓ runTemplateCLI > logs a link when template.about.repository is defined 12300ms
- ✓  bingo  src/preparation/prepareOptions.test.ts (6 tests) 12026ms
     - ✓ prepareOptions > prepare > uses an option value from produce when settings do not have options 640ms
     - ✓ prepareOptions > prepare > uses an option value from produce when settings do not have the options value 607ms
     - ✓ prepareOptions > prepare > uses an option value from settings when settings have the options value 603ms
     - ✓ prepareOptions > prepare > prepares options when given a template 10170ms

</details>

astrochemx committed:
test(bingo): update `runInput` test to handle platform-specific directory commands

astrochemx committed:
test(bingo-fs): normalize expected paths in `intake` test for cross-platform compatibility

JoshuaKGoldberg commented on packages/bingo-fs/src/intake.test.ts:
Now that `intake.ts` isn't being modified, these tests can be reverted, right?

astrochemx replied:
Sorry, I didn’t fully understand the question, as we haven't modified `intake.ts`, as far as I remember.

However, I made these changes because the `intake.ts` tests were failing on Windows due to incorrect slash usage in the expected paths – forward slashes (`/`) on POSIX and backward slashes (`\`) on Windows.

JoshuaKGoldberg commented on packages/bingo/src/cli/getTemplatePackageData.test.ts:
```diff
@@ -18,7 +18,52 @@ vi.mock("read-package-up", () => ({
+					? nodePathWindows.dirname(path)
+					: nodePathPosix.dirname(path),
+		},
+	};
```
[Testing] Refactoring nitpick, we can shrink this to be a bit less syntax-heavy:

```suggestion
	const nodePathWindows =
		await vi.importActual<typeof import("node:path/win32")>("node:path/win32");
	const nodePathPosix =
		await vi.importActual<typeof import("node:path/posix")>("node:path/posix");
	return {
		default: {
			...(await vi.importActual("node:path")),
			dirname: (path: string) =>
				isWindowsPaths
					? nodePathWindows.dirname(path)
					: nodePathPosix.dirname(path),
		},
	};
```

JoshuaKGoldberg reviewed (changes requested):
Looking great! No more comments from me on the implementation. Just a couple notes on tests.

astrochemx committed:
test(bingo): small refactoring of `getTemplatePackageData.test.ts`

Co-authored-by: Josh Goldberg ✨ <git@joshuakgoldberg.com>

JoshuaKGoldberg reviewed (approved):
Awesome, thanks so much for iterating on this so much! This is a good fix and I'm excited for Windows users to be unbroken.

![Shere Khan in The Jungle Book (1967) slowly clapping "Bravo, bravo, an extraordinary performance"](https://github.com/user-attachments/assets/c3518308-6869-4cbf-8df8-9f9e117656ee)
